### PR TITLE
DocC setup

### PR DESCRIPTION
**Pull request by yim-lee**

### CODE REVIEW

**Comment by ktoso** [comment]
We figured out a slightly better way to deal with the package files -- could you make files for "old swift versions" and just keep one for the new ones as `Package.swift`? This way we don't have to keep adding new files for every new Swift release in the future. See our doc about the migration for details.

File: Sources/StatsdClient/Documentation.docc/index.md [hunk 1]
```diff
@@ -0,0 +1,49 @@
+# ``StatsdClient``
```
Comment (tomerd) [inline Sources/StatsdClient/Documentation.docc/index.md]: I saw some of the NIO PRs are use `Docs.docc/index.md` and here and in distributed actors we use `Documentation.docc/index.md`. no strong opinion but I would prefer to get them consistent. which one do we prefer

cc @Lukasa @weissi @yim-lee @ktoso?

File: Sources/StatsdClient/Documentation.docc/index.md [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+* Recorder -> Histogram
+* Timer -> Timer
+                                              
+## Topics
+                                              
+### Metrics API
+
+- ``StatsdClient/makeCounter(label:dimensions:)``
+- ``StatsdClient/makeRecorder(label:dimensions:aggregate:)``
+- ``StatsdClient/makeTimer(label:dimensions:)``
```
Comment (tomerd) [inline Sources/StatsdClient/Documentation.docc/index.md]: do we need to adda few more here?

File: scripts/soundness.sh [hunk 3]
```diff
@@ -67,14 +73,14 @@ for language in swift-or-c bash dtrace; do
   matching_files=( -name '*' )
   case "$language" in
       swift-or-c)
-        exceptions=( -name c_nio_http_parser.c -o -name c_nio_http_parser.h -o -name cpp_magic.h -o -name Package.swift -o -name CNIOSHA1.h -o -name c_nio_sha1.c -o -name ifaddrs-android.c -o -name ifaddrs-android.h)
+        exceptions=( -name c_nio_http_parser.c -o -name c_nio_http_parser.h -o -name cpp_magic.h -o -name Package.swift -o -name Package@swift-5.6.swift -o -name CNIOSHA1.h -o -name c_nio_sha1.c -o -name ifaddrs-android.c -o -name ifaddrs-android.h)
```
Comment (tomerd) [inline scripts/soundness.sh]: `Package@swift-5.6.swift` -> `Package@*.swift`

Comment (tomerd) [reply scripts/soundness.sh]: also not sure we need al this nio files, maybe copy pasta?

File: Sources/StatsdClient/Documentation.docc/index.md (hunk 1 above)
Comment (yim-lee) [reply Sources/StatsdClient/Documentation.docc/index.md]: It's my PRs that are not being consistent so let's stick with `Docs.docc`.

File: scripts/soundness.sh (hunk 3 above)
Comment (yim-lee) [reply scripts/soundness.sh]: amended https://github.com/apple/swift-statsd-client/pull/36/commits/b112adbf796ebe160377d6d8921c773d3e8030ab

File: Sources/StatsdClient/Documentation.docc/index.md (hunk 1 above)
Comment (ktoso) [reply Sources/StatsdClient/Documentation.docc/index.md]: Let's go with `Docs` I'll adjust others

File: Sources/StatsdClient/Documentation.docc/index.md (hunk 2 above)
Comment (yim-lee) [reply Sources/StatsdClient/Documentation.docc/index.md]: added `init` and `shutdown`. do you have others in mind?

**Review by ktoso — APPROVED** [review]
LGTM looks good to me now!

File: Sources/StatsdClient/Documentation.docc/index.md (hunk 1 above)
Comment (ktoso) [reply Sources/StatsdClient/Documentation.docc/index.md]: We have `Docs.docc` actually https://github.com/apple/swift-distributed-actors/tree/main/Sources/DistributedActors/Docs.docc :) al let's stick to that everywhere 👍

File: Sources/StatsdClient/Documentation.docc/index.md (hunk 2 above)
Comment (tomerd) [reply Sources/StatsdClient/Documentation.docc/index.md]: `makeGauge` also?

**Comment by tomerd** [comment]
I think the "trick" with the `Package.swift` is not actually the right way. @fabianfett and myself went through this with the lambda-runtime and iirc it break downs in some scenarios. @fabianfett do you recall the details?

File: Sources/StatsdClient/Documentation.docc/index.md (hunk 2 above)
Comment (yim-lee) [reply Sources/StatsdClient/Documentation.docc/index.md]: i don't see `makeGauge`?

Comment (tomerd) [reply Sources/StatsdClient/Documentation.docc/index.md]: I guess this API is all coming from swift-metrics, its a protocol implementation

**Comment by Lukasa** [comment]
I don't think we're playing a trick here. 

Concretely, the observed behaviour of SwiftPM around the `@swift-x.y` package manifests is utterly baffling. SwiftPM seems to have a heuristic that combines the version in the filename with the tools version in order to select what manifest to use. However, this proposal clearly does work: if the tools version in `Package.swift` is _newer_ than the tools version in any other manifest, SwiftPM will prefer it for future releases.

**Comment by tomerd** [comment]
here is the history of going back and forth on this with aws-lambda-runtime: https://github.com/swift-server/swift-aws-lambda-runtime/pull/254#issuecomment-1104435274

iirc things worked okay locally and in CI, but broke when consuming the package as a dependency which is why we rolled it back. @fabianfett may recall more details

**Comment by Lukasa** [comment]
The breakage appears to be the result of using symlinks, which we aren't doing elsewhere.

**Comment by Lukasa** [comment]
Concretely, the documentation _explains_ a behaviour, but that behaviour demonstrably diverges from what SwiftPM actually does. Notice that the documentation has _both_ filenames _and_ tools versions. The tools versions appear to bear more of the burden than the filenames.

File: Sources/StatsdClient/Documentation.docc/index.md (hunk 2 above)
Comment (ktoso) [reply Sources/StatsdClient/Documentation.docc/index.md]: There isn't a makeGauge but Gauge calls into makeRecorder (for better or worse that's how it is)

**Comment by ktoso** [comment]
Hmmm, I can confirm that the "comment version" takes precedence over what the filename says, I've seen that.

I can also confirm having a `Package.swift` with tools version 5.6 is picked up correctly by a 5.7 swift, so that all seems to work like we'd want to  -- and what the current pattern carters to (so old versions with @ and new one without).

**Comment by tomerd** [comment]
@Lukasa we agree the way this is documented and works is confusing. @abertelrud @neonichu could you chime in here on the original design, and what is the recommended approach?

in any case, I think we should wait to hear back from @fabianfett, since he rolled back the "old version files" style from https://github.com/swift-server/swift-aws-lambda-runtime when we added 5.6 support there, and I cannot remember why - only that some scenario was not working for him.

**Comment by tomerd** [comment]
@yim-lee @ktoso @Lukasa until we hear more back from @fabianfett, let go with the "old files style", i.e make Package.swift required 5.6 and add files for each one of the old version we support.

**Review by tomerd — APPROVED** [review]

**Pull request merged**